File: CSharp/OpenTraceability/Interfaces/IEvent.cs
using OpenTraceability.Models.Common;
using OpenTraceability.Models.Events;
using OpenTraceability.Models.Events.KDEs;
using OpenTraceability.Models.Identifiers;
using OpenTraceability.Utility.Attributes;
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace OpenTraceability.Interfaces
{
    /// <summary>
    /// Interface for all events.
    /// </summary>
    public interface IEvent
    {
        Uri? EventID { get; set; }
        string? CertificationInfo { get; set; }
        DateTimeOffset? EventTime { get; set; }
        TimeSpan? EventTimeZoneOffset { get; set; }
        DateTimeOffset? RecordTime { get; set; }
        EventType EventType { get; }
        EventAction? Action { get; set; }
        Uri? BusinessStep { get; set; }
        Uri? Disposition { get; set; }
        PersistentDisposition? PersistentDisposition { get; set; }
        ErrorDeclaration? ErrorDeclaration { get; set; }
        EventLocation? Location { get; set; }
        EventReadPoint? ReadPoint { get; set; }
        List<EventBusinessTransaction> BizTransactionList { get; set; }
        List<EventSource> SourceList { get; set; }
        List<EventDestination> DestinationList { get; set; }
        List<IEventKDE> KDEs { get; }
        List<SensorElement> SensorElementList { get; set; }
        ReadOnlyCollection<EventProduct> Products { get; }
        T? GetKDE<T>(string ns, string name) where T : IEventKDE;
        T? GetKDE<T>() where T : IEventKDE;
        void AddProduct(EventProduct product);
        CertificationList? CertificationList { get; set; }
        EventILMD? GetILMD();
    }
}

File: Java/OpenTraceability/src/main/opentraceability/interfaces/ievent.kt
package interfaces

import models.events.kdes.CertificationList
import models.events.*
import java.time.Duration
import java.net.URI
import java.time.OffsetDateTime

interface IEvent {

    var EventID: URI?
    var CertificationInfo: String?
    var EventTime: OffsetDateTime?
    var EventTimeZoneOffset: Duration?
    var RecordTime: OffsetDateTime?
    var EventType: EventType?
    var Action: EventAction?
    var BusinessStep: URI?
    var Disposition: URI?
    var PersistentDisposition: PersistentDisposition?
    var ErrorDeclaration: ErrorDeclaration?
    var Location: EventLocation
    var ReadPoint: EventReadPoint
    var BizTransactionList: List<EventBusinessTransaction>
    var SourceList: List<EventSource>
    var DestinationList: List<EventDestination>
    var KDEs: List<IEventKDE>
    var SensorElementList: List<SensorElement>
    var Products: List<EventProduct>
    var CertificationList: CertificationList?

    fun <T> GetKDE(ns: String, name: String): T
    fun <T> GetKDE(): T
    fun AddProduct(product: EventProduct): Void
    fun GetILMD(): EventILMD?
}
